docs/tutorial/part_4.rst

```diff
@@ -0,0 +1,162 @@
+                
+                self._switch_to_window(0)
+                self._post_message('hello')
+                WebDriverWait(self.driver, 2).until(lambda _:
+                    'hello' in self._chat_log_value,
+                    'Message was not received by window 1 from window 1')
+                
+                self._switch_to_window(1)
+                self._post_message('world')
+                WebDriverWait(self.driver, 2).until(lambda _:
+                    'world' in self._chat_log_value,
+                    'Message was not received by window 2 from window 2')
+                self.assertTrue('hello' not in self._chat_log_value,
+                    'Message was improperly received by window 2 from window 1')
+            finally:
+                self._close_all_new_windows()
+        
+        # === Utility ===
+        
+        def _enter_chat_room(self, room_name):
+            self.driver.get(self.live_server_url + '/chat/')
+            ActionChains(self.driver).send_keys(room_name + '\n').perform()
+            WebDriverWait(self.driver, 2).until(lambda _:
+                room_name in self.driver.current_url)
+        
+        def _open_new_window(self):
+            self.driver.execute_script('window.open("about:blank", "_blank");')
+            self.driver.switch_to_window(self.driver.window_handles[-1])
+        
+        def _close_all_new_windows(self):
+            while len(self.driver.window_handles) > 1:
+                self.driver.switch_to_window(self.driver.window_handles[-1])
+                self.driver.execute_script('window.close();')
+            if len(self.driver.window_handles) == 1:
+                self.driver.switch_to_window(self.driver.window_handles[0])
+        
+        def _switch_to_window(self, window_index):
+            self.driver.switch_to_window(self.driver.window_handles[window_index])
+        
+        def _post_message(self, message):
+            ActionChains(self.driver).send_keys(message + '\n').perform()
+        
+        @property
+        def _chat_log_value(self):
+            return self.driver.find_element_by_css_selector('#chat-log').get_property('value')
+
+Our test suite extends ``ChannelsLiveServerTestCase`` rather than Django's usual
+suites for end-to-end tests (``StaticLiveServerTestCase`` or ``LiveServerTestCase``) so
+that URLs inside the Channels routing configuration like ``/ws/room/ROOM_NAME/``
+will work inside the suite.
+
+To run the tests, run the following command::
+
+    $ python3 manage.py test chat.tests
+
+You should see output that looks like::
+
+    Creating test database for alias 'default'...
+    System check identified no issues (0 silenced).
+    ..
+    ----------------------------------------------------------------------
+    Ran 2 tests in 5.014s
+    
+    OK
+    Destroying test database for alias 'default'...
+
+You now have a tested chat server!
+
+This is the end of the tutorial.
```

Comment:
Let's make this a bit more celebratory and link off to some other documentation they might want to read next